storage/src/adb/current.rs

```diff
@@ -0,0 +1,556 @@
+
+            // Delete that one key.
+            db.delete(&mut hasher, k1).await.unwrap();
+            db.commit(&mut hasher).await.unwrap();
+            assert_eq!(db.op_count(), 6); // 1 update, 2 commits, 2 moves, 1 delete.
+            let root2 = db.root(&mut hasher);
+            db.close().await.unwrap();
+            db = open_db(context.clone(), &mut hasher, partition).await;
+            assert_eq!(db.op_count(), 6); // 1 update, 2 commits, 2 moves, 1 delete.
+            assert_eq!(db.root(&mut hasher), root2);
+
+            // Confirm all activity bits are false
+            for i in 0..db.op_count() {
+                assert!(!db.status.get_bit(i));
+            }
+        });
+    }
+
+    /// Apply random operations to the given db, committing them (randomly & at the end) only if
+    /// `commit_changes` is true.
+    async fn apply_random_ops<E: RStorage + Clock + Metrics>(
+        hasher: &mut Sha256,
+        num_elements: u64,
+        commit_changes: bool,
+        rng_seed: u64,
+        db: &mut Current<E, Digest, Digest, Sha256, TwoCap, 64>,
+    ) -> Result<(), Error> {
+        // Log the seed with high visibility to make failures reproducible.
+        warn!("rng_seed={}", rng_seed);
+        let mut rng = StdRng::seed_from_u64(rng_seed);
+
+        for i in 0u64..num_elements {
+            let k = hash(&i.to_be_bytes());
+            let v = hash(&rng.next_u32().to_be_bytes());
+            db.update(hasher, k, v).await.unwrap();
+        }
+
+        // Randomly update / delete them. We use a delete frequency that is 1/7th of the update
+        // frequency.
+        for _ in 0u64..num_elements * 10 {
+            let rand_key = hash(&(rng.next_u64() % num_elements).to_be_bytes());
+            if rng.next_u32() % 7 == 0 {
+                db.delete(hasher, rand_key).await.unwrap();
+                continue;
+            }
+            let v = hash(&rng.next_u32().to_be_bytes());
+            db.update(hasher, rand_key, v).await.unwrap();
+            if commit_changes && rng.next_u32() % 20 == 0 {
+                // Commit every ~20 updates.
+                db.commit(hasher).await.unwrap();
+            }
+        }
+        if commit_changes {
+            db.commit(hasher).await.unwrap();
+        }
+
+        Ok(())
+    }
+
+    /// This test builds a random database, and makes sure that its state is correctly restored
+    /// after closing and re-opening.
+    #[test_traced("WARN")]
+    pub fn test_current_db_build_random_close_reopen() {
+        // Number of elements to initially insert into the db.
+        const ELEMENTS: u64 = 1000;
+
+        let executor = deterministic::Runner::default();
+        executor.start(|context| async move {
+            let mut hasher = Sha256::new();
+            let partition = "build_random";
+            // Use a non-deterministic rng seed to ensure each run is different.
+            let rng_seed = OsRng.next_u64();
```

Comment:
You can provide the seed during runtime init: https://github.com/commonwarexyz/monorepo/blob/103caa03ef469272b29f9f89a35bf27bd2d8f8a9/consensus/src/simplex/mod.rs#L1570

This will allow you to just use `context` 👍

Reply:
We should IMO avoid running each CI with different seeds unless done explicitly over a range (otherwise we'll just end up with flaking)

Reply:
(Once we standup our simulation cluster, we can run these over a range of seeds continuously)

Reply:
changed to use the context RNG to get a seed.